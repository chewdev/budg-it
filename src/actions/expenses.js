import {
  database,
  ref,
  get,
  child,
  push,
  set,
  remove,
  update,
} from "../firebase/firebase";

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    const userExpensesRef = ref(database, `users/${uid}/expenses`);
    const newExpenseRef = push(userExpensesRef);

    return set(newExpenseRef, expense).then((ref) => {
      dispatch(
        addExpense({
          id: newExpenseRef.key,
          ...expense,
        })
      );
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const expenseToRemoveRef = ref(database, `users/${uid}/expenses/${id}`);
    return remove(expenseToRemoveRef).then(() => {
      dispatch(
        removeExpense({
          id,
        })
      );
    });
  };
};

export const removeExpensesChosen = (ids = []) => ({
  type: "REMOVE_EXPENSES_CHOSEN",
  ids,
});

export const startRemoveExpensesChosen = (ids = []) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    ids.forEach((id) => {
      const expenseToRemoveRef = ref(database, `users/${uid}/expenses/${id}`);
      return remove(expenseToRemoveRef).then(() => {
        dispatch(removeExpense({ id }));
      });
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const expenseToUpdateRef = ref(database, `users/${uid}/expenses/${id}`);
    return update(expenseToUpdateRef, updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses,
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return get(child(ref(database), `users/${uid}/expenses`)).then(
      (snapshot) => {
        const expenses = [];

        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setExpenses(expenses));
      }
    );
  };
};
