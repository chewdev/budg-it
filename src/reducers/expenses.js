// Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id);
        case 'REMOVE_EXPENSES_CHOSEN':
            action.ids.forEach((id) => {
            state = state.filter((expense) => expense.id !== id);
            });
            return state;
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                };
            });
        case 'SET_EXPENSES':
            return action.expenses;
        case 'REMOVE_ALL':
            return [];
        default:
            return state;
    }
};