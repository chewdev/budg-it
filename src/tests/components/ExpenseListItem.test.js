import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import ExpenseListItem from '../../components/ExpenseListItem';

test('should render ExpenseListItem component with expense passed in', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
});
