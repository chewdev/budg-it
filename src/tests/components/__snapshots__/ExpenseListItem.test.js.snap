// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem component with expense passed in 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/expense/1"
>
  <div>
    <h3
      className="list-item__description"
    >
      Gum
    </h3>
    <span
      className="list-item__date"
    >
      December 31st, 1969
    </span>
  </div>
  <h3
    className="list-item__amount"
  >
    $1.95
  </h3>
</Link>
`;
